pageserver/src/remote_storage/storage_sync.rs

```diff
@@ -240,82 +376,22 @@ fn storage_sync_loop<P: std::fmt::Debug, S: 'static + RemoteStorage<StoragePath
             }
         })
         .cloned()
-        .map(SyncTask::UrgentDownload)
-        .collect::<Vec<_>>();
-    info!(
-        "Will download {} timelines to restore state",
-        urgent_downloads.len()
-    );
-    let mut accessor = SYNC_QUEUE.lock().unwrap();
-    accessor.extend(urgent_downloads.into_iter());
-    drop(accessor);
+        .map(SyncKind::Download)
+        .for_each(|sync_kind| SYNC_QUEUE.push(SyncTask::new(0, sync_kind)));
+}
 
-    while !crate::tenant_mgr::shutdown_requested() {
-        let mut queue_accessor = SYNC_QUEUE.lock().unwrap();
-        let next_task = queue_accessor.pop();
-        let remaining_queue_length = queue_accessor.len();
-        drop(queue_accessor);
+fn get_next_tasks(max_concurrent_sync: NonZeroUsize) -> Vec<SyncTask> {
```

Comment:
This is its own can of worms.

First, allowing default (100) concurrent upload/download tasks for uncompressed files feels wrong. That hopefully gets fixed soon in the archiving PR, so not super critical.

Then, in general, the tasks may overlap (there's nothing in the code that ensures they don't) and cause unneeded races.

For now, it's fine to pick tasks in batch due to incremental nature of all checkpoints and the fact that downloads are either for entire timeline or for nothing: neither uploads and downloads intersect now.

I think it's fine to leave it as is for now, later there will be more fun with archive defragmentation, where we could think on batching more, possibly add some heuristics to merge the tasks.

But for now, when thinking of remote files as archives, it seems wasteful to not to batch multiple tasks.